concordium-consensus/src/Concordium/GlobalState/Persistent/ReleaseSchedule.hs

```diff
@@ -0,0 +1,415 @@
+
+    addAccountRelease ts addr br = do
+        LegacyReleaseSchedule{..} <- refLoad br
+        refMake $!
+            LegacyReleaseSchedule
+                { rsFirstTimestamp = min rsFirstTimestamp ts,
+                  rsMap = Map.alter addAcc ts rsMap,
+                  rsEntryCount = 1 + rsEntryCount
+                }
+      where
+        addAcc Nothing = Just $! Set.singleton addr
+        addAcc (Just accs) = Just $! Set.insert addr accs
+
+    updateAccountRelease oldts ts addr br = do
+        LegacyReleaseSchedule{..} <- refLoad br
+        let remMap = Map.alter remAcc oldts rsMap
+        refMake $!
+            LegacyReleaseSchedule
+                { -- Since the new timestamp must be less than or equal to the old, this is correct.
+                  rsFirstTimestamp = min rsFirstTimestamp ts,
+                  rsMap = Map.alter addAcc ts remMap,
+                  rsEntryCount = rsEntryCount
+                }
+      where
+        remAcc Nothing = error "updateAccountRelease: no entry at expected release time"
+        remAcc (Just s) =
+            let s' = Set.delete addr s
+             in if Set.null s' then Nothing else Just s'
+        addAcc Nothing = Just (Set.singleton addr)
+        addAcc (Just accs) = Just $! Set.insert addr accs
+
+    processReleasesUntil ts br = do
+        rs@LegacyReleaseSchedule{..} <- refLoad br
+        if ts < rsFirstTimestamp
+            then return ([], br)
+            else do
+                let (newRS, accs) = go rs []
+                (accs,) <$> refMake newRS
+      where
+        go rs@LegacyReleaseSchedule{..} accum
+            | Map.null rsMap = (LegacyReleaseSchedule (Timestamp maxBound) Map.empty 0, accum)
+            | minTS <= ts =
+                go
+                    (rs{rsMap = newMap, rsEntryCount = rsEntryCount - fromIntegral (Set.size accs)})
+                    (accum ++ Set.toList accs)
+            | otherwise = (rs{rsFirstTimestamp = minTS}, accum)
+          where
+            -- Pattern match lazily in case the map is empty.
+            ~((minTS, accs), newMap) = Map.deleteFindMin rsMap
+
+-- |A set of accounts represented by 'AccountIndex'.
+newtype AccountSet = AccountSet {theAccountSet :: Set.Set AccountIndex}
+    deriving (Serialize)
+
+instance MonadBlobStore m => BlobStorable m AccountSet
+instance Applicative m => Cacheable m AccountSet
+
+-- |A release schedule for the P5 protocol version.  This uses a 'Trie.Trie' mapping 'Timestamp's
+-- to sets of accounts.
+data NewReleaseSchedule = NewReleaseSchedule
+    { -- |The first timestamp at which a release is scheduled (or the maximum possible timestamp if
+      -- no releases are scheduled).
+      rs5FirstTimestamp :: !Timestamp,
+      -- |A map recording the first release time for each account with a pending release.
+      -- An account should occur at most once in the map.
+      rs5Map :: !(Trie.TrieN BufferedFix Timestamp AccountSet)
```

Comment:
I'm guessing this relies quite crucially on the fact timestamps are serialized in big-endian. Could we please record this explicitly?

Reply:
It does not. We do require the instance `OrdFixedTrieKey Timestamp`, which is derived via `Word64`, and does rely on big-endian serialization for `Word64`.

Reply:
I'm not sure why you say it does not, but then in the followup sentence you say that it does. In particular the findMin operation requires this. I understand that there is a OrdFixedTrieKey abstraction, but it would help so much to understand this if we just explicitly stated that using the Trie here makes sense because of big endian serialziation.

Reply:
It does not rely on the `Serialize Timestamp` instance. Documenting this here would be like documenting why the `Ord` instance is correct for the key type in any structure that uses `Map`, which seems out of place to me.